resources/api/reservation.py

```diff
@@ -508,6 +509,66 @@ def has_object_permission(self, request, view, obj):
         return obj.can_modify(request.user)
 
 
+
+class ReservationAuthenticationLevelPermission(permissions.BasePermission):
+    """
+    This class matches the authentication level on resource with the current
+    authentication backend used to authenticate the user. User authenticated with
+    higher level authentication can reserve resource with lower level authentication,
+    plus the same level as user's authentication level.
+
+    User's auth level       Reserveable resource with auth level
+      Strong          ->       strong, mid, weak, none
+      Mid             ->       mid, weak, none
+      Weak            ->       weak, none
+    """
+    message = ''
+    # TODO: Needs discussing with client which login method belongs to which level.
+    STRONG_AUTHENTICATION = ('suomifi')
+    MID_AUTHENTICATION = ('axiell_aurora')
+    WEAK_AUTHENTICATION = ('google', 'github', 'facebook')
+
+    def has_permission(self, request, view):
+        resource_id = request.data.get('resource')
+        if request.method in permissions.SAFE_METHODS or not resource_id:
+            return True
+
+        resource = Resource.objects.get(id=resource_id)
+        return self._can_reserve_resource_with_current_login(request, resource)
+
+    def _can_reserve_resource_with_current_login(self, request, resource):
+        resource_authentication = resource.authentication
+        user_authentication = get_user_auth_backend(request)
+
+        if resource_authentication == 'none':
+            return True
+
+        if user_authentication in self.STRONG_AUTHENTICATION:
+            return True
+
+        if (
+            user_authentication in self.MID_AUTHENTICATION
+            and resource_authentication == 'strong'
+        ):
+            self.message = _('You need to login with strong authentication to reserve '
+                             'this resource.')
+            return False
+
+        if (
+            user_authentication in self.WEAK_AUTHENTICATION
+            and resource_authentication != 'weak'
+        ):
+            if resource_authentication == 'mid':
+                self.message = _('You need to login with mid authentication to reserve '
+                                 'this resource.')
+            else:
+                self.message = _('You need to login with strong authentication to '
+                                 'reserve this resource.')
+            return False
+
+        return True
```

Comment:
I think a good practise with permission checks is to default to `False` and only deviate from that if a specific condition applied. That makes sure you're not giving permission in some edge case that you just hadn't thought of and the if statements don't catch.

```suggestion
    def _can_reserve_resource_with_current_login(self, request, resource):
            resource_authentication = resource.authentication
            user_authentication = get_user_auth_backend(request)
    
            if resource_authentication == 'none':
                return True
    
            if user_authentication in self.STRONG_AUTHENTICATION:
                return True
    
            if user_authentication in self.MID_AUTHENTICATION:
                if resource_authentication in ["mid", "weak"]:
                    return True
                else:
                    self.message = _('You need to login with strong authentication to reserve '
                                    'this resource.')
    
            if user_authentication in self.WEAK_AUTHENTICATION:
                if resource_authentication == "weak":
                    return True
                elif resource_authentication == 'mid':
                    self.message = _('You need to login with mid authentication to reserve '
                                     'this resource.')
                else:
                    self.message = _('You need to login with strong authentication to '
                                     'reserve this resource.')
    
            return False
```

Reply:
Eh, don't know why Github doesn't like to show the diff of the suggestions in a nicer way.